### PR TITLE
[Infer Symbolic Shape ops.yaml No.130] Add cholesky solve

### PR DESCRIPTION
**Pull request by ykcai-daniel**
### PR Category
CINN

### PR Types
 Improvements

### Description
Add cholesky solve infer meta.
Add a function to compute the batch dimension with broadcast

### CODE REVIEW

**Comment by ykdcai** [comment]
The logic of shape inference for triangular solve, choleskey solve and solve are actually the same. Maybe we can combine them into one infermeta?

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc [hunk 1]
```diff
@@ -313,12 +313,78 @@ bool BmmOpInferSymbolicShape(pir::Operation *op,
...
+    batchPortion[i] = x_size != 1 ? x_size : y_size;
+  }
+  return batchPortion;
+}
+
+bool CholeskySolveOpInferSymbolicShape(pir::Operation *op,
+                                       pir::InferSymbolicShapeContext
+                                       *infer_context) {
+  const std::vector<DimExpr> &x_dims =
+      infer_context->GetShapeOrDataForValue(op->operand_source(0)).shape();
```
Comment (gongshaotian) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc:352]: 命名不规范，shape dialect中有shape区和data区的设计，这里推荐使用x_shape

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc [hunk 2]
```diff
@@ -313,12 +313,78 @@ bool BmmOpInferSymbolicShape(pir::Operation *op,
...
+  const std::vector<DimExpr> &y_dims =
+      infer_context->GetShapeOrDataForValue(op->operand_source(1)).shape();
```
Comment (gongshaotian) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc:354]: 同上

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc [hunk 3]
```diff
@@ -313,12 +313,78 @@ bool BmmOpInferSymbolicShape(pir::Operation *op,
...
+  
+  auto x_dims_n = x_dims.size();
+  auto y_dims_n = y_dims.size();
```
Comment (gongshaotian) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc]: 非必要不使用auto，推荐const size_t

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc [hunk 4]
```diff
@@ -313,12 +313,78 @@ bool BmmOpInferSymbolicShape(pir::Operation *op,
...
+
+  PADDLE_ENFORCE_GE(x_dims_n,
+                    2,
+                    common::errors::InvalidArgument(
+                        "the rank of input Y must greater or equal to 2"));
+  PADDLE_ENFORCE_GE(y_dims_n,
```
Comment (gongshaotian) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc]: 这两个断言可以合并成一个

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc [hunk 5]
```diff
@@ -313,12 +313,78 @@ bool BmmOpInferSymbolicShape(pir::Operation *op,
...
+                    common::errors::InvalidArgument(
+                        "the rank of input X must greater or equal to 2"));
+  
+  // input Matrix Y should be square matrix
+  infer_context->AddEqualCstr(y_dims[y_dims_n - 1], y_dims[y_dims_n - 2]);
+  // the first dim of Matrix X must be equal to the first dim of Matrix Y
+  infer_context->AddEqualCstr(x_dims[x_dims_n - 2],  y_dims[y_dims_n - 2]);
+
+  // Output size is batch size+ last two dims of x
+  std::vector<DimExpr> x_batch_dims(x_dims.begin(), x_dims.end() - 2);
```
Comment (gongshaotian) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc]: 添个namespace: symbol::DimExpr

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc [hunk 6]
```diff
@@ -313,12 +313,78 @@ bool BmmOpInferSymbolicShape(pir::Operation *op,
...
-// }
+
+static inline std::vector<DimExpr> MatrixGetBroadcastBatchPortion(
+    const std::vector<DimExpr>& x, const std::vector<DimExpr>& y) {
+  size_t size_x = x.size();
+  size_t size_y = y.size();
+  size_t size = std::max(size_x, size_y);
+  std::vector<DimExpr> batchPortion(size);
+
+  ptrdiff_t i = (ptrdiff_t)size - 1;
```
Comment (gongshaotian) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc]: 这里直接用int就行

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc [hunk 7]
```diff
@@ -313,12 +313,78 @@ bool BmmOpInferSymbolicShape(pir::Operation *op,
...
+        (x_size == y_size || x_size == 1 || y_size == 1),
+        true,
+        common::errors::PreconditionNotMet(
+            "The size of tensor x (%d) must match the size of tensor y "
+            "(%d) at non-singleton dimension %d.",
+            x_size,
+            y_size,
+            i));
+
+    batchPortion[i] = x_size != 1 ? x_size : y_size;
```
Comment (gongshaotian) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc]: for 循环内部使用 builder.Broadcast() 实现两个维度的广播，并添加broadcast的约束。可参考MatmulOpInferSymbolicShape

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc [hunk 8]
```diff
@@ -313,12 +313,78 @@ bool BmmOpInferSymbolicShape(pir::Operation *op,
...
+  std::vector<DimExpr> x_batch_dims(x_dims.begin(), x_dims.end() - 2);
+  std::vector<DimExpr> y_batch_dims(y_dims.begin(), y_dims.end() - 2);
+
+  // Batch dimensions should be broadcastable
+  // Get batch size, which is possibly broadcasted
+  std::vector<DimExpr> expand_batch_portion =
+      MatrixGetBroadcastBatchPortion(x_batch_dims, y_batch_dims);
+
+  // append the size of x to batch dimentions
+  std::vector<DimExpr> result({expand_batch_portion});
```
Comment (gongshaotian) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc]: result命名不规范

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc [hunk 9]
```diff
@@ -313,12 +313,82 @@ bool BmmOpInferSymbolicShape(pir::Operation *op,
...
 }
 
-// bool CholeskySolveOpInferSymbolicShape(pir::Operation *op,
-//                                        pir::InferSymbolicShapeContext
-//                                        *infer_context) {
-//   // pass
-//   return true;
-// }
+// Given two batch dimensions x and y, return the broadcasted batch dimension of
+// x and y
```
Comment (gongshaotian) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc]: 删掉非必要注释，下同

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc [hunk 10]
```diff
@@ -313,12 +313,82 @@ bool BmmOpInferSymbolicShape(pir::Operation *op,
...
+
+  // Output size is batch size+ last two dims of x
+  std::vector<symbol::DimExpr> x_batch_dims(x_shape.begin(), x_shape.end() - 2);
+  std::vector<symbol::DimExpr> y_batch_dims(y_shape.begin(), y_shape.end() - 2);
+
+  // Batch dimensions should be broadcastable
+  // Get batch size, which is possibly broadcasted
+  PADDLE_ENFORCE_EQ(
+      (x_batch_dims.size() == 0 && y_batch_dims.size != 0) ||
+          (x_batch_dims.size() != 0 && y_batch_dims.size == 0),
```
Comment (gongshaotian) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc]: 这个enforce可以删掉

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc [hunk 11]
```diff
@@ -313,12 +313,82 @@ bool BmmOpInferSymbolicShape(pir::Operation *op,
...
+      false,
+      common::errors::InvalidArgument(
+          "One of x and y have batch dimension yet the other do not."));
+  std::vector<symbol::DimExpr> expand_batch_portion =
+      MatrixGetBroadcastBatchPortion(x_batch_dims, y_batch_dims, infer_context);
+
+  // append the size of x to batch dimentions
+  std::vector<symbol::DimExpr> output_shape({expand_batch_portion});
```
Comment (gongshaotian) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc]: 多包了一层vector

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc [hunk 12]
```diff
@@ -313,12 +313,82 @@ bool BmmOpInferSymbolicShape(pir::Operation *op,
...
-// }
+// Given two batch dimensions x and y, return the broadcasted batch dimension of
+// x and y
+static inline std::vector<symbol::DimExpr> MatrixGetBroadcastBatchPortion(
+    const std::vector<symbol::DimExpr> &x,
+    const std::vector<symbol::DimExpr> &y,
+    pir::InferSymbolicShapeContext *infer_context) {
+  size_t size_x = x.size();
+  size_t size_y = y.size();
+  size_t size = std::max(size_x, size_y);
```
Comment (gongshaotian) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc]: 命名成max_size吧

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc [hunk 13]
```diff
@@ -313,12 +313,82 @@ bool BmmOpInferSymbolicShape(pir::Operation *op,
...
+  for (int i = std::static_cast<int>(size) - 1; i >= 0; --i) {
+    int offset = size - i - 1;
+    int dim_x = size_x - offset - 1;
+    int dim_y = size_y - offset - 1;
+
+    if (dim_x < 0) {
+      batchPortion[i] = y[dim_y];
+    } else if (dim_y < 0) {
+      batchPortion[i] = x[dim_x];
+    } else if (dim_x >= 0 && dim_y >= 0) {
```
Comment (gongshaotian) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc]: 这么写有点不够直观，可以分成两阶段，伪代码：
`
int diff = x.size() -y.size();
if (diff > 0){
  for(i=0; i < diff; i++){
    broadcast_out = x[i];
  }
}else{
   ...
    .... = y[i];
}

for (i == diff ; i < max_size; i++ ){
    ...->addBroadcastcstr()
   = builder.broadcast();
}


`

**Comment by gongshaotian** [comment]
注意变量命名，尽量直观且语义完备无歧义；本地编译、测试(/vdb1/yubaoku/paddle_src/Paddle/test/legacy_test/test_cholesky_op.py)通过后再推上来跑CI

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc [hunk 14]
```diff
@@ -313,12 +313,75 @@ bool BmmOpInferSymbolicShape(pir::Operation *op,
...
+    }
+  } else {
+    size_diff = -size_diff;
+    for (int i = 0; i < size_diff; i++) {
+      batchPortion[i] = y[i];
+    }
+  }
+
+  symbol::DimExprBuilder builder;
+  for (int i = diff; i < max_size; i++) {
```
Comment (gongshaotian) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc]: diff未定义

Comment (ykcai-daniel) [reply paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc]: 已修改

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc [hunk 15]
```diff
@@ -313,12 +313,75 @@ bool BmmOpInferSymbolicShape(pir::Operation *op,
...
+    pir::InferSymbolicShapeContext *infer_context) {
+  // use int to avoid underflow for minus
+  int size_x = x.size();
+  int size_y = y.size();
+  int max_size = std::max(size_x, size_y);
+  std::vector<symbol::DimExpr> batchPortion(max_size);
+
+  int size_diff = size_x - size_y;
+  if (size_diff > 0) {
+    for (int i = 0; i < diff; i++) {
```
Comment (gongshaotian) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc:328]: 这里还有一个diff没有删掉

**Pull request closed without merging**